src/IdentityModel/Client/AccessTokenHandler.cs

```diff
@@ -0,0 +1,195 @@
+// Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.
+
+using IdentityModel.Internal;
+using System;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace IdentityModel.Client
+{
+    /// <summary>
+    /// HTTP message handler that encapsulates access token handling and renew
+    /// </summary>
+    public class AccessTokenHandler : DelegatingHandler
+    {
+        private readonly SemaphoreSlim _lock = new SemaphoreSlim(1, 1);
+        private readonly TokenClient _tokenClient;
+        private readonly string _scope;
+
+        private string _accessToken;
+        private bool _disposed;
+
+        /// <summary>
+        /// Gets or sets the timeout
+        /// </summary>
+        public TimeSpan Timeout { get; set; } = TimeSpan.FromSeconds(5);
+        
+        /// <summary>
+        /// Gets the current access token
+        /// </summary>
+        public string AccessToken
+        {
+            get
+            {
+                if (_lock.Wait(Timeout))
+                {
+                    try
+                    {
+                        return _accessToken;
+                    }
+                    finally
+                    {
+                        _lock.Release();
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Occurs when the tokens were renewed successfully
+        /// </summary>
+        public event EventHandler<TokenRenewedEventArgs> TokenRenewed = delegate { };
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RefreshTokenHandler"/> class.
+        /// </summary>
+        /// <param name="tokenEndpoint">The token endpoint.</param>
+        /// <param name="clientId">The client identifier.</param>
+        /// <param name="clientSecret">The client secret.</param>
+        /// <param name="scope">The scope.</param>
+        /// <param name="accessToken">The access token.</param>
+        /// <param name="innerHandler">The inner handler.</param>
+        public AccessTokenHandler(string tokenEndpoint, string clientId, string clientSecret, string scope, string accessToken = null, HttpMessageHandler innerHandler = null)
+            : this(new TokenClient(tokenEndpoint, clientId, clientSecret), scope, accessToken, innerHandler)
+        { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RefreshTokenHandler"/> class.
+        /// </summary>
+        /// <param name="client">The client.</param>
+        /// <param name="scope">The scope.</param>
+        /// <param name="accessToken">The access token.</param>
+        /// <param name="innerHandler">The inner handler.</param>
+        public AccessTokenHandler(TokenClient client, string scope, string accessToken = null, HttpMessageHandler innerHandler = null)
```

Comment:
Would an overload that could take a `DiscoveryClient` be desirable? Might make less setup work for consumers.